scripts: dtlib/edtlib: Make output consistent on Python 3.5-

ulfalizer opened the pull request:
Change:

```
scripts: dtlib/edtlib: Make output consistent on Python 3.5-

Python 3.5 and earlier do not preserve dictionary insertion order when
iterating over dictionaries, and do not give the same order between
runs. This broke the dtlib and edtlib test suites and made the output
jump around randomly between runs. It also made device INST_<n> numbers
non-deterministic, which broke some code on Python 3.5 (though
hardcoding device instance numbers in the code might be a bit shaky).

Fix it by using collections.OrderedDict instead of plain dict wherever
order matters. This makes the output identical on all supported Python
versions. It also allows testdtlib.py and testedtlib.py to run in CI,
which uses Python 3.5.

Fixes: #20571
```

Semi-related cleanup:

```
scripts: edtlib: Avoid modifying the global yaml.(C)Loader

edtlib is a library, and modifying yaml.(C)Loader directly interferes
with any binding loading in edtlib clients. To avoid that, add a custom
loader for bindings.

Internally, PyYAML does this, which is why defining a separate class
works:

    @classmethod
    def add_constructor(cls, tag, constructor):
        if not 'yaml_constructors' in cls.__dict__:
            cls.yaml_constructors = cls.yaml_constructors.copy()
        cls.yaml_constructors[tag] = constructor
```

Small cleanup:

```
scripts: edtlib: Move imports after overview comment

Might make the comment easier to spot.
```

carlescufi commented:
@ulfalizer I assume a `Fixes:` is missing for: https://github.com/zephyrproject-rtos/zephyr/issues/20571 ?

pabigot reviewed (commented):
Unless the YAML and dictionary changes are tightly coupled those should have been in separate commits, starting with the YAML one as cleanup, then the dictionary one which needs the custom loader.  Piling everything into one big diff makes it a pain to review.

galak reviewed (commented):
I'm with @pabigot that I'd like to see this as 2 commits, but the changes themselves look good.

ulfalizer committed:
scripts: edtlib: Avoid modifying the global yaml.(C)Loader

edtlib is a library, and modifying yaml.(C)Loader directly interferes
with any binding loading in edtlib clients. To avoid that, add a custom
loader for bindings.

Internally, PyYAML does this, which is why defining a separate class
works:

    @classmethod
    def add_constructor(cls, tag, constructor):
        if not 'yaml_constructors' in cls.__dict__:
            cls.yaml_constructors = cls.yaml_constructors.copy()
        cls.yaml_constructors[tag] = constructor

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
scripts: dtlib/edtlib: Make output consistent on Python 3.5-

Python 3.5 and earlier do not preserve dictionary insertion order when
iterating over dictionaries, and do not give the same order between
runs. This broke the dtlib and edtlib test suites and made the output
jump around randomly between runs. It also made device INST_<n> numbers
non-deterministic, which broke some code on Python 3.5 (though
hardcoding device instance numbers in the code might be a bit shaky).

Fix it by using collections.OrderedDict instead of plain dict wherever
order matters. This makes the output identical on all supported Python
versions. It also allows testdtlib.py and testedtlib.py to run in CI,
which uses Python 3.5.

Fixes: #20571

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer committed:
scripts: edtlib: Move imports after overview comment

Might make the comment easier to spot.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer commented:
Split up now, along with some more details.

pabigot reviewed (approved):
Thanks; much easier to understand now.
